Add user management e2e tests

sbryzak opened the pull request:
Tests for https://github.com/codeready-toolchain/host-operator/pull/142

sbryzak commented:
/retest

sbryzak commented:
/retest

sbryzak commented:
/retest

MatousJobanek commented:
/retest

MatousJobanek commented on test/e2e/user_management_test.go:
```diff
@@ -42,7 +45,99 @@ func (s *userManagementTestSuite) TearDownTest() {
+		s.setApprovalPolicyConfig("automatic")
+
+		// then
+		s.checkUserBanned()
```
could you please replace this call of the `checkUserBanned` method by the actual subtests? The current state is confusing.

sbryzak replied:
Certainly, I was just copying the structure in `usersignup_test` which does exactly this, however I've got no problem putting the subtests into one single test method.

MatousJobanek replied:
Thanks, it looks much better now :+1:

MatousJobanek commented on test/e2e/user_management_test.go:
```diff
@@ -99,3 +194,27 @@ func (s *userManagementTestSuite) TestUserDisabled() {
+	return s.createAndCheckUserSignup(specApproved, "testuser"+id, "testuser"+id+"@test.com", approvedAutomatically()...)
+}
+
+func newBannedUser(host *wait.HostAwaitility, email string) *v1alpha1.BannedUser {
```
could you please move this function to `base_user_test.go` file? it's called only there

sbryzak replied:
Sure, however I should also point out that there's an equivalent `newUserSignup` function in `usersignup_test`.

MatousJobanek replied:
but `newUserSignup` is used inside of `usersignup_test`, so it's fine having it there

anyway, thanks for moving the `newBannedUser` function

MatousJobanek commented on wait/host.go:
```diff
@@ -126,6 +129,33 @@ func (a *HostAwaitility) WaitForUserSignup(name string, criteria ...UserSignupWa
+	// Ignore the error, as this implementation cannot return one
+	_, _ = md5hash.Write([]byte(email))
+	emailHash := hex.EncodeToString(md5hash.Sum(nil))
+	labels := map[string]string{toolchainv1alpha1.BannedUserEmailHashLabelKey: emailHash}
```
wouldn't it be good to extract this part into a separated function that would be called here as well as at these two places?:
https://github.com/codeready-toolchain/toolchain-e2e/pull/80/files#diff-cc443128665ec194918207e25d44b3eeR173
https://github.com/codeready-toolchain/toolchain-e2e/pull/80/files#diff-4f14cf49a084497bcfac1b29c3bbf79bR204

sbryzak replied:
Good idea, I've added a function for this.

MatousJobanek commented on test/e2e/base_user_test.go:
```diff
@@ -0,0 +1,125 @@
+	s.T().Logf("BannedUser '%s' created", bannedUser.Spec.Email)
+
+	bannedUser, err = s.hostAwait.WaitForBannedUser(email)
+	require.NoError(s.T(), err)
```
Let me understand one thing - you create a `BannedUser` and then you wait for the same `BannedUser` being present in the cluster. What is the purpose of the check/wait there? There is no prod logic tested here... You can just create the `BannedUser` and that's all.

sbryzak replied:
Good catch, I must have overlooked this.  Fixed!

MatousJobanek commented on test/e2e/usersignup_test.go:
```diff
@@ -189,82 +165,29 @@ func (s *userSignupIntegrationTest) checkUserSignupManualApproval() {
+	md5hash := md5.New()
+	_, _ = md5hash.Write([]byte(email))
+	emailHash := hex.EncodeToString(md5hash.Sum(nil))
+
```
could we have one unified place where we create all `UserSignup` instances? ideally using the same way as is used from the registration-service: https://github.com/codeready-toolchain/toolchain-e2e/blob/master/test/e2e/e2e_test.go#L255-L271

sbryzak replied:
You want to create all `UserSignup` instances via the registration service API?  We can't actually do this for the banned user test where the user is banned prior to user registration, as the registration service now queries for any matching `BannedUser` resources before allowing the registration to proceed.  This would break some of our tests.  It seems to me that `newUserSignup` is already used extensively for creating `UserSignup` instances.  In fact, after checking all of `base_user_test` and `user_management_test` code it seems this is the *only* place where the instance is created.

MatousJobanek replied:
I already mentioned that it's possible covering the other cases too, but keep it as it is :+1:

MatousJobanek commented on test/e2e/user_management_test.go:
```diff
@@ -42,7 +45,99 @@ func (s *userManagementTestSuite) TearDownTest() {
+
+	err = s.hostAwait.WaitUntilMasterUserRecordDeleted(mur.Name)
+	require.NoError(s.T(), err)
+}
```
Could you please also test a case when the user is reactivated?

sbryzak replied:
Sure thing, I've added a test for this.

MatousJobanek replied:
thanks, but it would be great if you could add it as a subtest of the `ban provisioned usersignup` test - it would be much easier to read

alexeykazakov replied:
Hm.. why not as a subtest for the deactivation test as is?

MatousJobanek replied:
because this is easier to read and undrestand what is the target of the tests:
```go
s.T().Run("ban provisioned usersignup", func(t *testing.T) {
	s.setApprovalPolicyConfig("automatic")

	// Create a new UserSignup and confirm it was approved automatically
	userSignup, mur := s.createUserSignupAndAssertAutoApproval(false)

	// Create the BannedUser
	bannedUser := s.createAndCheckBannedUser(userSignup.Annotations[v1alpha1.UserSignupUserEmailAnnotationKey])

	// Confirm the user is banned
	_, err := s.hostAwait.WithRetryOptions(wait.TimeoutOption(time.Second*10)).WaitForUserSignup(userSignup.Name,
		wait.UntilUserSignupHasConditions(approvedAutomaticallyAndBanned()...))
	require.NoError(s.T(), err)

	// Confirm that a MasterUserRecord is deleted
	_, err = s.hostAwait.WithRetryOptions(wait.TimeoutOption(time.Second * 10)).WaitForMasterUserRecord(userSignup.Spec.Username)
	require.Error(s.T(), err)

	t.Run("unban the banned user", func(t *testing.T) {
		// Unban the user
		err = s.awaitility.Client.Delete(context.TODO(), bannedUser)
		require.NoError(s.T(), err)

		// Confirm the BannedUser is deleted
		err = s.hostAwait.WaitUntilBannedUserDeleted(bannedUser.Name)
		require.NoError(s.T(), err)

		// Confirm the user is provisioned
		_, err = s.hostAwait.WithRetryOptions(wait.TimeoutOption(time.Second*10)).WaitForUserSignup(userSignup.Name,
			wait.UntilUserSignupHasConditions(approvedAutomatically()...))
		require.NoError(s.T(), err)

		// Confirm the MUR is created
		_, err = s.hostAwait.WaitForMasterUserRecord(mur.Name)
		require.NoError(s.T(), err)
	})
})
```
just my point of view ;-)

alexeykazakov replied:
OK. You are refereeing to different tests :)
You comment was added to (and mentioned) the deactivation/re-activation tests. But you actually meant ban/unban tests. Now it makes sense.
Here we go: https://github.com/codeready-toolchain/toolchain-e2e/pull/80/commits/61dcbceaf564b1e3b70190332c9f052e94a77aaa

MatousJobanek replied:
wrong copy-paste with the name of the test :man_shrugging: sorry for that
and thanks for changing it :-)

MatousJobanek commented on test/e2e/user_management_test.go:
```diff
@@ -42,7 +45,99 @@ func (s *userManagementTestSuite) TearDownTest() {
+		email := "testuser" + id + "@test.com"
+		s.createAndCheckBannedUser(email)
+
+		_ = s.createAndCheckUserSignupNoMUR(false, "testuser"+id, email, banned()...)
```
here, you only check that the UserSignup exists with the given condition, but not that the MUR doesn't exist - you should add that check too

sbryzak replied:
That's not so simple... the MUR isn't ever created in the first place, and so a compliant name is never generated for it and therefore I don't even have an MUR name to use for the check.  Do you have any ideas for how this can be easily checked?

MatousJobanek replied:
yes, by listing all MURs. 
You can also use the label selector for the `toolchain.dev.openshift.com/user-id` label equal to `UserSignup.name`

alexeykazakov replied:
Actually, in this particular case the username is enough to check the MUR. It's a regular username (testuser + UUID), so it would match the compliant name.
I've added the check in https://github.com/codeready-toolchain/toolchain-e2e/pull/80/commits/47aff7e47cb17aa3ae92efe2064202c4b3837775

MatousJobanek commented on test/e2e/user_management_test.go:
```diff
@@ -42,7 +45,99 @@ func (s *userManagementTestSuite) TearDownTest() {
+		body, err := ioutil.ReadAll(resp.Body)
+		require.NoError(s.T(), err)
+		require.NotNil(s.T(), body)
+		assert.Equal(s.T(), http.StatusInternalServerError, resp.StatusCode)
```
Is it also possible to check the content of the error?

sbryzak replied:
Done

MatousJobanek commented on test/e2e/user_management_test.go:
```diff
@@ -42,7 +45,99 @@ func (s *userManagementTestSuite) TearDownTest() {
+
+func (s *userManagementTestSuite) TestUserBanning() {
+
+	s.T().Run("user banning", func(t *testing.T) {
```
could you please also add a case when:
1. user was provisioned
2. user got banned
3. the ban was removed
4. user is provisioned again

sbryzak replied:
Done

sbryzak commented on test/e2e/base_user_test.go:
```diff
@@ -0,0 +1,125 @@
+	userSignup.Spec.Approved = specApproved
+	err := s.awaitility.Client.Create(context.TODO(), userSignup, testsupport.CleanupOptions(s.testCtx))
+	require.NoError(s.T(), err)
+	s.T().Logf("user signup '%s' created", userSignup.Name)
```
No we can't use `createAndApproveSignup` here, as it initiates a signup process via the registration service with a predetermined approval progression - i.e. it first checks that the signup matches the `pendingApproval()` conditions, then after a manual approval checks that the conditions match `approvedByAdmin()`.  

The `createAndCheckUserSignupNoMUR()` function allows the expected conditions to be passed in, which in the case of the `createAndCheckUserSignupNoMUR` test is the `banned()` conditions, so `createAndApproveSignup` will fail here.

MatousJobanek replied:
> The createAndCheckUserSignupNoMUR() function allows the expected conditions to be passed in, which in the case of the create usersignup with preexisting banneduser test is the banned() conditions, so createAndApproveSignup will fail here.

if you want to verify this scenario (that the controller won't create the MUR based on the UserSignup) then with a small modification of `createAndApproveSignup` function it should be possible:

1. call `createAndApproveSignup` but don't approve the signup yet
2. create BannedUser
3. approve the signup

but I understand that it's slightly hacky.

MatousJobanek replied:
As I mentioned below, I understand that it's easier to use it in this way - keep it as it is

sbryzak commented on test/e2e/base_user_test.go:
```diff
@@ -0,0 +1,125 @@
+	userSignup := s.createAndCheckUserSignupNoMUR(specApproved, username, email, conditions...)
+
+	// Confirm the MUR was created
+	mur := s.assertCreatedMUR(userSignup)
```
The `assertCreatedMUR()` function already calls `s.hostAwait.WaitForMasterUserRecord` to ensure that the MUR is created.  I could call `verifyResourcesProvisionedForSignup` however this seems like overkill, considering that `createAndCheckUserSignup` is only used indirectly by two user management tests.  Furthermore, confirming _all_ of the resources are provisioned is seemingly redundant in light of the nature of the tests which themselves are limited in their scope to just the `UserSignup` and `MasterUserRecord` resources.  I think we already have overly sufficient coverage of resource provisioning in other integration tests, wouldn't you agree?

MatousJobanek replied:
it's not about verifying the provisioning flow (yes, I agree the it's sufficiently covered), it's about making sure that the account is provisioned and ready before it will be banned. In other words - all resources are properly created before they will be deleted.

alexeykazakov replied:
Done in https://github.com/codeready-toolchain/toolchain-e2e/pull/80/commits/47aff7e47cb17aa3ae92efe2064202c4b3837775

alexeykazakov replied:
Done in https://github.com/codeready-toolchain/toolchain-e2e/pull/80/commits/944f6941d8fbf701d9ae8fdb4e4f61733aa7b8e5

sbryzak commented on go.mod:
```diff
@@ -1,14 +1,15 @@
+	github.com/codeready-toolchain/api v0.0.0-20200305152414-a958daa81263
+	github.com/codeready-toolchain/host-operator v0.0.0-20200306134032-ba201c72fe2b // indirect
+	github.com/codeready-toolchain/member-operator v0.0.0-20200305072445-1691a6259bfe // indirect
+	github.com/codeready-toolchain/registration-service v0.0.0-20200305025344-33697732d256 // indirect
```
I do have a dependency for registration service as I'm now checking the user signup response and unmarshaling the json into one of the registration service error types.  I'll try and work out what the other ones are for.

sbryzak replied:
I've removed both the host-operator and member-operator dependencies and it seems there is no negative effect.

MatousJobanek replied:
could you please run `go mod tidy`?

alexeykazakov replied:
Done in https://github.com/codeready-toolchain/toolchain-e2e/pull/80/commits/47aff7e47cb17aa3ae92efe2064202c4b3837775

sbryzak commented:
@MatousJobanek thanks for the great review! I've hopefully now addressed all your comments, could you please check over the PR again and please approve if satisfied with the changes.  That way I can get both this PR and https://github.com/codeready-toolchain/host-operator/pull/142 merged over the weekend.

sbryzak commented:
/retest

alexeykazakov commented:
/retest

alexeykazakov commented:
/retest

MatousJobanek commented:
> @MatousJobanek thanks for the great review! I've hopefully now addressed all your comments, could you please check over the PR again and please approve if satisfied with the changes. That way I can get both this PR and codeready-toolchain/host-operator#142 merged over the weekend.

@sbryzak sorry for not making the review on Friday - I had to end earlier because of family reasons and when you wrote the message I was just logging off.

alexeykazakov commented on wait/host.go:
```diff
@@ -46,13 +47,30 @@ func (a *HostAwaitility) WaitForMasterUserRecord(name string, criteria ...Master
 	return mur, err
 }
 
+func (a *HostAwaitility) GetMasterUserRecord(criteria ...MasterUserRecordWaitCriterion) (*toolchainv1alpha1.MasterUserRecord, error) {
```
We do not always need Wait. Sometime we really need to use pure Get. In this particular case we use Get to make sure that MUR does not exist. Using Wait would slow down the test.
In other cases we could need to check that some object already exists vs. will be created shortly.

MatousJobanek replied:
ok, make sense. I missed it that it's used for the case when MUR is not present. :+1:

alexeykazakov commented on test/e2e/e2e_test.go:
```diff
@@ -142,7 +142,7 @@ func TestE2EFlow(t *testing.T) {
 		mur, err := hostAwaitility.WaitForMasterUserRecord(johnsmithName)
 		require.NoError(t, err)
-		teamRevisions, err := getRevisions(awaitility, "team", "dev", "stage")
+		teamRevisions, err := getTeamTierRevisions(awaitility)
```
I was hopping I can merge this PR first so you can deal with the conflicts :grin:


Just kidding: https://github.com/codeready-toolchain/toolchain-e2e/pull/80/commits/22ac4f16ddade9d1f7f0ff5d5c284dc1d3085ff3
